### PR TITLE
EXOGW-360: OpenTelemetry Support

### PR DESCRIPTION
**Pull request by steve-keep**
- [x] Trace Base Resolver Functions
- [x] Trace Hooks
- [x] Trace Mikro Provider
- [x] Trace Rest Provider
- [x] Trace initial Request details - Post Body payload
- [x] Add Lambda event ID to main span attribute
- [x] Trace Custom Queries and Mutations
- [x] Trace Field Resolvers

## Summary by CodeRabbit

- **New Features**
  - Introduced tracing functionality to improve tracking and debugging across the application.

- **Updates**
  - Updated version numbers for multiple packages to 1.0.1 to reflect recent changes and improvements.

### CODE REVIEW

**Comment by steve-keep** [comment]
@coderabbitai review

File: src/packages/mikroorm/src/provider/provider.ts [hunk 1]
```diff
@@ -231,11 +233,15 @@ export class MikroBackendProvider<D> implements BackendProvider<D> {
 		return collectedPaths;
 	};
 
+	@TraceMethod()
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/mikroorm/src/provider/provider.ts:236]: The span is ended inside the decorator and so this is not required.

File: src/packages/rest/src/provider/provider.ts [hunk 2]
```diff
@@ -71,12 +81,15 @@ export class RestBackendProvider<D = unknown> implements Provider<D> {
...
+	@TraceMethod()
 	public async findByRelatedId(
 		entity: unknown,
 		relatedField: string,
 		relatedFieldIds: string[],
-		filter?: Filter<D>
+		filter?: Filter<D>,
+		trace?: Trace
 	): Promise<D[]> {
+		trace?.span.updateName(`Rest - findByRelatedId`);
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/rest/src/provider/provider.ts:92]: This is handled in the decorator

File: src/examples/xero/src/backend/schema/profit-and-loss-row.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ import {
 
 const defaultSort: Record<string, Sort> = { ['date']: Sort.DESC };
 
-const parseReport = (tenantId: string, report: ReportWithRows) => {
+const parseReport = async (tenantId: string, report: ReportWithRows) => {
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/examples/xero/src/backend/schema/profit-and-loss-row.ts:20]: removed async as not needed

File: src/packages/mikroorm/src/provider/provider.ts [hunk 4]
```diff
@@ -349,8 +361,10 @@
 
 	public async updateOne(
 		id: string | number,
-		updateArgs: Partial<D & { version?: number }>
+		updateArgs: Partial<D & { version?: number }>,
+		trace?: Trace
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/mikroorm/src/provider/provider.ts:365]: done 👍

File: src/packages/core/src/types.ts [hunk 5]
```diff
@@ -92,11 +94,7 @@ export interface BackendProvider<D> {
 
 	entityType?: new () => D;
 
-	find(
-		filter: Filter<D>,
-		pagination?: PaginationOptions,
-		additionalOptionsForBackend?: any
-	): Promise<D[]>;
+	find(filter: Filter<D>, pagination?: PaginationOptions, trace?: Trace): Promise<D[]>;
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/core/src/types.ts:97]: `trace` is an optional param.

File: src/packages/core/src/field-resolver.ts [hunk 6]
```diff
@@ -42,10 +44,15 @@ export const fieldResolver = (
 				entityMetadata: metadata,
 				fieldMetadata: relationship,
 			});
+			span?.end();
```
Comment (thekevinbrown) [inline src/packages/core/src/field-resolver.ts:47]: Do we need to `try {} finally {}` here to make sure we end the span?

File: src/packages/core/src/query-manager.ts [hunk 7]
```diff
@@ -98,15 +99,20 @@ const visit = async <D>(
...
-		entityName,
-		filter,
-		pagination,
-	}: {
-		entityName: string;
-		filter?: Filter<D>;
-		pagination?: PaginationOptions;
-	}) => {
+	@TraceMethod()
+	async find<G = unknown, D = unknown>(
```
Comment (thekevinbrown) [inline src/packages/core/src/query-manager.ts:103]: 👨‍🍳 💋

File: src/packages/core/src/resolvers.ts [hunk 8]
```diff
@@ -144,7 +156,11 @@ export const list = async <G, D>({
 	});
 	logger.trace({ result }, 'Got result');
 
+	const fromBackendEntitySpan = trace?.tracer.startSpan(
+		`FromBackendEntity - ${result.length} ${entity.name} ${result.length > 1 ? 'entities' : 'entity'}.`
+	);
 	result = result.map((resultRow) => fromBackendEntity(entity, resultRow));
+	fromBackendEntitySpan?.end();
```
Comment (thekevinbrown) [inline src/packages/core/src/resolvers.ts:163]: Do we need to `try {} finally {}` here to make sure we end the span?

File: src/packages/core/src/field-resolver.ts [hunk 9]
```diff
@@ -42,10 +44,15 @@ export const fieldResolver = (
...
+			return res;
 		}
 
 		if (typeof property === 'function') {
-			return property(source, args, context, info);
+			const span = tracer?.startSpan(`FieldResolver - ${parent}.${key} - Function`);
+			const res = property(source, args, context, info);
+			span?.end();
```
Comment (thekevinbrown) [inline src/packages/core/src/field-resolver.ts:54]: Do we need to `try {} finally {}` here to make sure we end the span?

File: src/packages/mikroorm/src/provider/provider.ts [hunk 10]
```diff
@@ -248,7 +254,9 @@ export class MikroBackendProvider<D> implements BackendProvider<D> {
 		//     id
 		//   }
 		// }
+		const filterConversionSpan = trace?.tracer.startSpan('Convert filter to Mikro-Orm format');
 		const where = filter ? gqlToMikro(JSON.parse(JSON.stringify(filter))) : undefined;
+		filterConversionSpan?.end();
```
Comment (thekevinbrown) [inline src/packages/mikroorm/src/provider/provider.ts:259]: Do we need to `try {} finally {}` here to make sure we end the span?

File: src/packages/mikroorm/src/provider/provider.ts [hunk 11]
```diff
@@ -280,12 +288,11 @@ export class MikroBackendProvider<D> implements BackendProvider<D> {
...
 		query.populate((driver as any).autoJoinOneToOneOwner(meta, []));
 
-		if (additionalOptionsForBackend?.populate) {
-			query.populate(additionalOptionsForBackend.populate);
-		}
-
 		try {
+			const getResultSpan = trace?.tracer.startSpan('Mikro-Orm - Fetch Data');
 			const result = await query.getResult();
+			getResultSpan?.end();
```
Comment (thekevinbrown) [inline src/packages/mikroorm/src/provider/provider.ts:294]: Do we need to `try {} finally {}` here to make sure we end the span?

**Review by thekevinbrown — APPROVED** [review]
Love the pattern. Only real question is anywhere that we're doing a:

```
beginTrace()
...do stuff
endTrace()
```
sort of pattern, we should probably `try` / `finally` that so that the trace is always ended.

We could also provide a helper that takes a callback to do this for us so it looks like

```typescript
traceOperation('name', async () => {
     doStuff
});
```

if that feels cleaner.

**Comment by steve-keep** [comment]
> Love the pattern. Only real question is anywhere that we're doing a:
> 
> ```
> beginTrace()
> ...do stuff
> endTrace()
> ```
> 
> sort of pattern, we should probably `try` / `finally` that so that the trace is always ended.
> 
> We could also provide a helper that takes a callback to do this for us so it looks like
> 
> ```ts
> traceOperation('name', async () => {
>      doStuff
> });
> ```
> 
> if that feels cleaner.

Great spot thanks! Wrapped them all in the `trace()` or `traceSync()` functions!

**Pull request merged**